plugins/c8y_configuration_plugin/src/download.rs

```diff
@@ -0,0 +1,299 @@
+async fn download_config_file(
+    plugin_config: &PluginConfig,
+    smartrest_request: SmartRestConfigDownloadRequest,
+    tedge_config: &TEdgeConfig,
+    http_client: &mut JwtAuthHttpProxy,
+) -> Result<(), anyhow::Error> {
+    // Convert smartrest request to config download request struct
+    let mut config_download_request =
+        ConfigDownloadRequest::try_new(smartrest_request, plugin_config, tedge_config)?;
+
+    // Confirm that the file has write access before any http request attempt
+    let () = config_download_request.has_write_access()?;
+
+    // If the provided url is c8y, add auth
+    if http_client.url_is_in_my_tenant_domain(config_download_request.download_info.url()) {
+        let token = http_client.get_jwt_token().await?;
+        config_download_request.download_info.auth = Some(Auth::new_bearer(&token.token()));
+    }
+
+    // Download a file to tmp dir
+    let downloader = config_download_request.create_downloader();
+    let () = downloader
+        .download(&config_download_request.download_info)
+        .await?;
+
+    // Move the downloaded file to the final destination
+    let () = config_download_request.move_file()?;
+
+    Ok(())
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct ConfigDownloadRequest {
+    pub download_info: DownloadInfo,
+    pub destination_path: PathBuf,
+    pub tmp_dir: PathBuf,
+    pub file_name: String,
+}
+
+impl ConfigDownloadRequest {
+    fn try_new(
+        request: SmartRestConfigDownloadRequest,
+        plugin_config: &PluginConfig,
+        tedge_config: &TEdgeConfig,
+    ) -> Result<Self, ConfigDownloadError> {
+        // Check if the requested config type is in the plugin config list
+        if !plugin_config.files.contains(&request.config_type) {
+            return Err(ConfigDownloadError::InvalidRequestedConfigType {
+                path: request.config_type,
+            });
+        }
+
+        let destination_path = PathBuf::from(request.config_type);
+        let tmp_dir = tedge_config.query(TmpPathDefaultSetting)?.into();
+        let file_name = Self::get_filename(destination_path.clone())?;
+
+        Ok(Self {
+            download_info: DownloadInfo {
+                url: request.url,
+                auth: None,
+            },
+            destination_path,
+            tmp_dir,
+            file_name,
+        })
+    }
+
+    fn get_filename(path: PathBuf) -> Result<String, ConfigDownloadError> {
+        let filename = path
+            .file_name()
+            .ok_or_else(|| ConfigDownloadError::FileNameNotFound { path: path.clone() })?
+            .to_str()
+            .ok_or_else(|| ConfigDownloadError::InvalidFileName { path: path.clone() })?
+            .to_string();
+        Ok(filename)
+    }
+
+    fn has_write_access(&self) -> Result<(), ConfigDownloadError> {
+        // The file does not exist before downloading a file
+        if !&self.destination_path.is_file() {
+            return Ok(());
```

Comment:
If the file doesn't exist, then the parent directory permission will have to be checked for write access, right? The file not existing doesn't guarantee write access, right?

Reply:
I would use an optimistic approach here. Try to write the file without any check and react only in case of an error.